### PR TITLE
Oracle: Fix duplicate identifier and streamline identifier

### PR DESCRIPTION
**Pull request by aimeos**
|      Q       |   A
|------------- | -----------
| Type         | bug/improvement
| BC Break     | yes
| Fixed issues | N/A

## Summary

The Oracle platform class creates duplicate identifier if the passed name is longer then 24 characters because the original code simply truncates everything beyond that limit.

Furthermore, it fixes upper/lower case of "_seq" and "_apk" suffix if the identifier is quoted to create e.g. "id_seq" instead of "id_SEQ" and uses the same format as the PostgreSQL platform for the generated sequence names.

### CODE REVIEW

**Comment by morozov** [comment]
There's no need to submit a patch for every release branch. We'll merge the original up once it's accepted.

File: src/Platforms/OraclePlatform.php [hunk 1]
```diff
@@ -522,11 +522,15 @@ private function addSuffix(string $identifier, string $suffix): string
      */
     private function getAutoincrementIdentifierName(Identifier $table): string
     {
-        $identifierName = $this->addSuffix($table->getName(), '_AI_PK');
+        $identifierName = $this->addSuffix($table->getName(), '_apk');
```
Comment (morozov) [inline src/Platforms/OraclePlatform.php:527]: Why is this necessary?

File: tests/Platforms/OraclePlatformTest.php [hunk 2]
```diff
@@ -563,10 +564,12 @@ public function testUsesSequenceEmulatedIdentityColumns(): void
...
     public function testReturnsIdentitySequenceName(): void
     {
-        self::assertSame('MYTABLE_SEQ', $this->platform->getIdentitySequenceName('mytable', 'mycolumn'));
-        self::assertSame('"mytable_SEQ"', $this->platform->getIdentitySequenceName('"mytable"', 'mycolumn'));
-        self::assertSame('MYTABLE_SEQ', $this->platform->getIdentitySequenceName('mytable', '"mycolumn"'));
-        self::assertSame('"mytable_SEQ"', $this->platform->getIdentitySequenceName('"mytable"', '"mycolumn"'));
+        self::assertSame('MYTABLE_MYCOLUMN_SEQ', $this->platform->getIdentitySequenceName('mytable', 'mycolumn'));
+        self::assertSame('"mytable_mycolumn_seq"', $this->platform->getIdentitySequenceName('"mytable"', 'mycolumn'));
+        self::assertSame('"mytable_mycolumn_seq"', $this->platform->getIdentitySequenceName('mytable', '"mycolumn"'));
+        self::assertSame('"mytable_mycolumn_seq"', $this->platform->getIdentitySequenceName('"mytable"', '"mycolumn"'));
```
Comment (morozov) [inline tests/Platforms/OraclePlatformTest.php:570]: Why is it required that the platform generates the same sequence name for three different tables/columns. Isn't the issue that the patch is trying to solve?

File: src/Platforms/OraclePlatform.php (hunk 1 above)
Comment (aimeos) [reply src/Platforms/OraclePlatform.php:527]: As you only have 30 chars for identfiers in Oracle, I tried to keep the suffix as short as possible - especially if the hash value must be added too. With that in mind, the `_AI_PK` suffix is just a waste of characters and I tried to use only four characters like for `_seq`. Thus, it gets less likely that the hash value must be added too.

File: tests/Platforms/OraclePlatformTest.php (hunk 2 above)
Comment (aimeos) [reply tests/Platforms/OraclePlatformTest.php:570]: The patch solves the issue of identifiers being too long. These tests are only for shorter identifiers depending on if they are quoted or not

File: src/Platforms/OraclePlatform.php (hunk 1 above)
Comment (morozov) [reply src/Platforms/OraclePlatform.php:527]: This isn't strictly necessary to address the issue in question. It requires the same testing as in https://github.com/doctrine/dbal/pull/5000#pullrequestreview-811419639.

File: tests/Platforms/OraclePlatformTest.php (hunk 2 above)
Comment (morozov) [reply tests/Platforms/OraclePlatformTest.php:570]: If an identifier is unquoted, it should be upper-cased before concatenation with another one or a suffix, and the result must be always quoted. This way, this logic will be the same as the logic on the server side. At least it looks like the original tests behave like this.

Comment (aimeos) [reply tests/Platforms/OraclePlatformTest.php:570]: Done. Names will now look like:
```
mytable_MYCOLUMN_seq
MYTABLE_mycolumn_seq
```

File: src/Platforms/OraclePlatform.php (hunk 1 above)
Comment (aimeos) [reply src/Platforms/OraclePlatform.php:527]: Similar to #5000:

Migration works as expected, (unique) index and foreign key names are not affected because the modified addSuffix() method is only used to generate sequence and autoincrement trigger names.

Existing sequences for tables with names change their name because column name and maybe the the hash characters are added to the sequence name. Trigger names for autoincrement columns are also dropped and recreated with a new name.

All sequence names change and that's breaking because the sequence get dropped and recreated. Therefore, the next ID starts with 1 and inserting new data into a table which already contains data will fail.

Comment (morozov) [reply src/Platforms/OraclePlatform.php:527]: > Therefore, the next ID starts with 1 and inserting new data into a table which already contains data will fail.

In order for this to be accepted even in a major release, an upgrade path should be provided. I.e. what should the affected API consumer do to handle this BC break on their side?

**Comment by derrabus** [comment]
Is this change meant to be a bugfix? In that case, 4.0.x is not the appropriate target branch.

File: src/Platforms/OraclePlatform.php [hunk 3]
```diff
@@ -841,12 +847,13 @@ public function usesSequenceEmulatedIdentityColumns(): bool
...
     public function getIdentitySequenceName(string $tableName, string $columnName): string
     {
-        $table = new Identifier($tableName);
+        $table = $this->normalizeIdentifier($tableName);
+        $column = $this->normalizeIdentifier($columnName);
 
-        // No usage of column name to preserve BC compatibility with <2.5
-        $identitySequenceName = $this->addSuffix($table->getName(), '_SEQ');
+        $colname = $columnName ? '_' . $column->getName() : '';
+        $identitySequenceName = $this->addSuffix($table->getName() . $colname, '_seq');
```
Comment (morozov) [inline src/Platforms/OraclePlatform.php:854]: Why does the column name need to be part of the sequence name? Can a table have multiple auto-increment columns?

Comment (aimeos) [reply src/Platforms/OraclePlatform.php:854]: Yes, tables can have multiple sequences in Oracle and PostgreSQL compared to only one autoincrement column in MySQL. That's also the reason why the DBAL PostgreSQL platform code creates sequences names as combination of table and column name.

**Comment by aimeos** [comment]
@derrabus This isn't only a bug fix. The PR includes a change in how sequence names are created to make them unique and this is highly breaking.
